### PR TITLE
[SPARK-15615][SQL] Add an API to load DataFrame from Dataset[String] storing JSON

### PR DESCRIPTION
**Pull request by pjfanning**
## What changes were proposed in this pull request?

SPARK-15615 proposes replacing the sqlContext.read.json(rdd) with a dataset equivalent.
SPARK-15463 adds a CSV API for reading from Dataset[String] so this keeps the API consistent.
I am deprecating the existing RDD based APIs.

## How was this patch tested?

There are existing tests. I left most tests to use the existing APIs as they delegate to the new json API.

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala [hunk 1]
```diff
@@ -328,18 +329,34 @@ class DataFrameReader private[sql](sparkSession: SparkSession) extends Logging {
    * @param jsonRDD input RDD with one JSON object per record
    * @since 1.4.0
    */
+  @deprecated("Use json(Dataset[String]) instead.", "2.2.0")
   def json(jsonRDD: RDD[String]): DataFrame = {
+    import sparkSession.sqlContext.implicits._
+    json(sparkSession.createDataset(jsonRDD))
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala]: nit `sparkSession.createDataset(jsonRDD)(Encoders.STRING)`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JsonFileFormat.scala [hunk 2]
```diff
@@ -122,11 +121,13 @@ class JsonFileFormat extends TextBasedFileFormat with DataSourceRegister {
...
     }
 
-    sparkSession.sparkContext.hadoopRDD(
+    val rdd = sparkSession.sparkContext.hadoopRDD(
       conf.asInstanceOf[JobConf],
       classOf[TextInputFormat],
       classOf[LongWritable],
       classOf[Text]).map(_._2.toString) // get the text line
+    import sparkSession.sqlContext.implicits._
+    sparkSession.createDataset(rdd)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JsonFileFormat.scala]: same here

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/TestJsonData.scala [hunk 3]
```diff
@@ -231,4 +231,10 @@ private[json] trait TestJsonData {
   lazy val singleRow: RDD[String] = spark.sparkContext.parallelize("""{"a":123}""" :: Nil)
 
   def empty: RDD[String] = spark.sparkContext.parallelize(Seq[String]())
+  
+  def dataset(rdd: RDD[String]): Dataset[String] = {
+    val sqlContext = spark.sqlContext
+    import sqlContext.implicits._
+    spark.createDataset(rdd)
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/TestJsonData.scala]: same here

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala [hunk 4]
```diff
@@ -1364,9 +1364,9 @@ class JsonSuite extends QueryTest with SharedSQLContext with TestJsonData {
     })
   }
 
-  test("SPARK-6245 JsonRDD.inferSchema on empty RDD") {
+  test("SPARK-6245 JsonRDD.inferSchema on empty Dataset") {
     // This is really a test that it doesn't throw an exception
-    val emptySchema = JsonInferSchema.infer(empty, "", new JSONOptions(Map.empty[String, String]))
+    val emptySchema = JsonInferSchema.infer(dataset(empty), "", new JSONOptions(Map.empty[String, String]))
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala]: I think we can just write `empty.toDS`

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/TestJsonData.scala [hunk 5]
```diff
@@ -231,4 +231,10 @@ private[json] trait TestJsonData {
   lazy val singleRow: RDD[String] = spark.sparkContext.parallelize("""{"a":123}""" :: Nil)
 
   def empty: RDD[String] = spark.sparkContext.parallelize(Seq[String]())
+  
+  def dataset(rdd: RDD[String]): Dataset[String] = {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/TestJsonData.scala]: actually we don't need this , see https://github.com/apache/spark/pull/16895/files#r100681578

**Comment by cloud-fan** [comment]
ok to test

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala [hunk 6]
```diff
@@ -1364,10 +1364,11 @@ class JsonSuite extends QueryTest with SharedSQLContext with TestJsonData {
     })
   }
 
-  test("SPARK-6245 JsonRDD.inferSchema on empty RDD") {
+  test("SPARK-6245 JsonRDD.inferSchema on empty Dataset") {
     // This is really a test that it doesn't throw an exception
+    val emptyDataset = spark.createDataset(empty)(Encoders.STRING)
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala]: doesn't `empty.toDS` work?

Comment (pjfanning) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala]: I can double check but the toDS call appears to require the spark implicits import

Comment (pjfanning) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala]: RDD[_] only has toDS() function added when SQLImplicits applies an implicit conversion to wrap the RDD as a DatasetHolder.
import sparkSession.sqlContext.implicits._

Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala]: I think the implicits are already imported at the beginning of this test suite

**Comment by cloud-fan** [comment]
LGTM, I'll merge it in 1 or 2 days, if no one agains this API change.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**